File: reader/src/main/java/com/freddieptf/reader/widgets/ReaderViewPager.java
package com.freddieptf.reader.widgets;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.viewpager.widget.ViewPager;

/**
 * Created by fred on 5/19/15.
 */
public class ReaderViewPager extends ViewPager {

    private float startDragXPos = 0;
    private ReadProgressListener readProgressListener;

    public ReaderViewPager(Context context) {
        this(context, null);
    }

    public ReaderViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setReadProgressListener(ReadProgressListener readProgressListener) {
        this.readProgressListener = readProgressListener;
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        float inDragX = ev.getX();
        switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN: {
                startDragXPos = ev.getX();
                break;
            }
            case MotionEvent.ACTION_MOVE:{
                if (startDragXPos < inDragX) {
                    // swipe from left to right
                    if(getCurrentItem() == 0) {
                        // if we are at the start
                        readProgressListener.onSwipeToPreviousCh();
                    }
                } else if(startDragXPos > inDragX) {
                    // swipe from right to left
                    if(getCurrentItem() == getAdapter().getCount()-1) {
                        // if we are at the end
                        readProgressListener.onSwipeToNextCh();
                    }
                }
                break;
            }
        }
        return super.onInterceptTouchEvent(ev);
    }

    public interface ReadProgressListener {
        void onSwipeToNextCh();
        void onSwipeToPreviousCh();
    }

}

File: reader/src/main/java/com/freddieptf/reader/widgets/ReaderViewPager.kt
package com.freddieptf.reader.widgets

import android.content.Context
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.MotionEvent
import androidx.core.view.GestureDetectorCompat
import androidx.viewpager.widget.ViewPager

/**
 * Created by fred on 5/19/15.
 */
class ReaderViewPager @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : ViewPager(context, attrs) {

    companion object {
        private val TAG = ReaderViewPager::class.java.simpleName
    }

    private var startDragXPos = 0f
    private var readProgressListener: ReadProgressListener? = null
    private var readSignals = ArrayList<SimpleReadSignals>()
    var direction: DIRECTION? = null
        private set
        get
    private var detector: GestureDetectorCompat? = null

    enum class DIRECTION {
        /**
         * LEFT TO RIGHT THE MANGA WAY
         **/
        LEFT_TO_RIGHT,
        /**
         * RIGHT TO LEFT THE NORMAL COMIC WAY
         * */
        RIGHT_TO_LEFT
    }

    fun setReadProgressListener(readProgressListener: ReadProgressListener) {
        this.readProgressListener = readProgressListener
    }

    fun setReadDirection(direction: DIRECTION) {
        this.direction = direction
        readSignals.forEach { it.onReadDirectionChange(this.direction!!) }
    }

    fun addReadSignalCallback(callbacks: SimpleReadSignals) {
        if (!readSignals.contains(callbacks)) readSignals.add(callbacks)
    }

    override fun setCurrentItem(item: Int, smoothScroll: Boolean) {
        var i = if (direction == DIRECTION.LEFT_TO_RIGHT) adapter!!.count - item - 1 else item
        super.setCurrentItem(i, smoothScroll)
    }

    override fun setCurrentItem(item: Int) {
        var i = if (direction == DIRECTION.LEFT_TO_RIGHT) adapter!!.count - item - 1 else item
        super.setCurrentItem(i)
    }

    override fun getCurrentItem(): Int {
        if (direction == DIRECTION.LEFT_TO_RIGHT) {
            return (adapter?.count ?: 0) - super.getCurrentItem() - 1
        }
        return super.getCurrentItem()
    }


    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        val inDragX = ev.x
        when (ev.action) {
            MotionEvent.ACTION_DOWN -> {
                startDragXPos = ev.x
            }
            MotionEvent.ACTION_MOVE -> {
                if (currentItem == 0) {
                    when {
                        (direction == DIRECTION.LEFT_TO_RIGHT && startDragXPos > inDragX) ||
                                (direction == DIRECTION.RIGHT_TO_LEFT && startDragXPos < inDragX) -> {
                            readProgressListener!!.onSwipeToPreviousCh()
                        }
                    }
                } else if (currentItem == (adapter?.count ?: 0) - 1) {
                    when {
                        (direction == DIRECTION.LEFT_TO_RIGHT && startDragXPos < inDragX) ||
                                (direction == DIRECTION.RIGHT_TO_LEFT && startDragXPos > inDragX) ->
                            readProgressListener!!.onSwipeToNextCh()
                    }
                }

            }
        }

        return super.onInterceptTouchEvent(ev)
    }

    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
        if (detector == null)
            detector = GestureDetectorCompat(context, gestureListenere)
        detector!!.onTouchEvent(ev)
        return super.dispatchTouchEvent(ev)
    }

    interface ReadProgressListener {
        fun onSwipeToNextCh()
        fun onSwipeToPreviousCh()
    }

    private interface ReadSignals {
        fun onPagerTapToFocus()
        fun onPageLongPress()
        fun onReadDirectionChange(direction: DIRECTION)
    }

    open class SimpleReadSignals : ReadSignals {
        override fun onPagerTapToFocus() {}
        override fun onReadDirectionChange(direction: DIRECTION) {}
        override fun onPageLongPress() {}
    }

    private var gestureListenere = object : GestureDetector.SimpleOnGestureListener() {
        override fun onSingleTapUp(e: MotionEvent?): Boolean {
            return true
        }

        override fun onSingleTapConfirmed(e: MotionEvent?): Boolean {
            readSignals.forEach { it.onPagerTapToFocus() }
            return true
        }

        override fun onLongPress(e: MotionEvent?) {
            readSignals.forEach { it.onPageLongPress() }
        }
    }

}
